src/rtc.rs

```diff
@@ -0,0 +1,448 @@
+        let hours = bcd2_decode(tr.ht().bits(), tr.hu().bits());
+        if self.is_24h_fmt() {
+            return Ok(rtcc::Hours::H24(hours as u8));
+        }
+        if !tr.pm().bit() {
+            return Ok(rtcc::Hours::AM(hours as u8));
+        }
+        Ok(rtcc::Hours::PM(hours as u8))
+    }
+
+    fn get_time(&mut self) -> Result<NaiveTime, Self::Error> {
+        self.set_24h_fmt();
+        let seconds = self.get_seconds().unwrap();
+        let minutes = self.get_minutes().unwrap();
+        let hours = hours_to_u8(self.get_hours().unwrap());
+
+        Ok(NaiveTime::from_hms(
+            hours.into(),
+            minutes.into(),
+            seconds.into(),
+        ))
+    }
+
+    fn get_weekday(&mut self) -> Result<u8, Self::Error> {
+        let dr = self.regs.dr.read();
+        let weekday = bcd2_decode(dr.wdu().bits(), 0x00);
+        Ok(weekday as u8)
+    }
+
+    fn get_day(&mut self) -> Result<u8, Self::Error> {
+        let dr = self.regs.dr.read();
+        let day = bcd2_decode(dr.dt().bits(), dr.du().bits());
+        Ok(day as u8)
+    }
+
+    fn get_month(&mut self) -> Result<u8, Self::Error> {
+        let dr = self.regs.dr.read();
+        let mt: u8 = if dr.mt().bit() { 1 } else { 0 };
+        let month = bcd2_decode(mt, dr.mu().bits());
+        Ok(month as u8)
+    }
+
+    fn get_year(&mut self) -> Result<u16, Self::Error> {
+        let dr = self.regs.dr.read();
+        let year = bcd2_decode(dr.yt().bits(), dr.yu().bits());
+        Ok(year as u16)
+    }
+
+    fn get_date(&mut self) -> Result<NaiveDate, Self::Error> {
+        let day = self.get_day().unwrap();
+        let month = self.get_month().unwrap();
+        let year = self.get_year().unwrap();
+
+        Ok(NaiveDate::from_ymd(year.into(), month.into(), day.into()))
+    }
+
+    fn get_datetime(&mut self) -> Result<NaiveDateTime, Self::Error> {
+        self.set_24h_fmt();
+
+        let day = self.get_day().unwrap();
+        let month = self.get_month().unwrap();
+        let year = self.get_year().unwrap();
+
+        let seconds = self.get_seconds().unwrap();
+        let minutes = self.get_minutes().unwrap();
+        let hours = hours_to_u8(self.get_hours().unwrap());
+
+        Ok(
+            NaiveDate::from_ymd(year.into(), month.into(), day.into()).and_hms(
+                hours.into(),
+                minutes.into(),
+                seconds.into(),
+            ),
+        )
+    }
+}
+
+// Two 32-bit registers (RTC_TR and RTC_DR) contain the seconds, minutes, hours (12- or 24-hour format), day (day
+// of week), date (day of month), month, and year, expressed in binary coded decimal format
+// (BCD). The sub-seconds value is also available in binary format.
+//
+// The following helper functions encode into BCD format from integer and
+// decode to an integer from a BCD value respectively.
+fn bcd2_encode(word: u32) -> (u8, u8) {
+    (
+        (word / 10).try_into().unwrap(),
+        (word % 10).try_into().unwrap(),
```

Comment:
I think instead of panicking we should return a `Result<(u8, u8), Error>` here with an `InvalidInputData` if the input is too large.

I think in the old PR I had advocated to making this check panicking, but I've since been convinced that it is better to limit panics in HAL code, as much as possible.